Enable CAs to be signed with Ed25519 private key for OpenSSH support

thinkwelltwd opened the pull request:
Here's a PR for enabling OpenSSH support. If this is outside your project scope, feel free to say so. Or if you have code suggestions, will do my best to implement them.

mathiasertl commented:
Hi,

I'm not against this, I'd like to merge this, with a few minor changes. 

But I'm also curious - what exactly is the use case here?

mathiasertl commented on ca/django_ca/managers.py:
```diff
@@ -336,9 +342,18 @@ def init(
         #       directly.
         generate_key_args = validate_key_parameters(key_size, key_type, ecc_curve)
-        algorithm = parse_hash_algorithm(algorithm)
+        algorithm = None if openssh_ca else parse_hash_algorithm(algorithm)
```
Please do not write this in a single line, as coverage tests will show this as covered even if (for example) `openssh_ca` is never `True`.

thinkwelltwd replied:
Converted to if statement

mathiasertl commented on ca/django_ca/models.py:
```diff
@@ -793,11 +793,6 @@ def key(self, password: Optional[Union[str, bytes]]) -> PrivateKeyTypes:
-        if isinstance(self._key, ed25519.Ed25519PrivateKey):  # pragma: nocover
-            raise ValueError("Ed25519 private keys are not supported.")
-        if isinstance(self._key, ed448.Ed448PrivateKey):  # pragma: nocover
-            raise ValueError("Ed25519 private keys are not supported.")
```
I think you should not remove the check for `ed448` keys, right?

Note that mypy checks are currently not in the GitHub CI pipeline, as it only works with the current HEAD of cryptography (in fact, I made quite a few PRs for this in cryptography). But there is a tox target that you can run locally for MyPy checks!

thinkwelltwd replied:
Restored ed448 check

mathiasertl commented on ca/django_ca/models.py:
```diff
@@ -1243,6 +1238,10 @@ def usable(self) -> bool:
 
+    @property
+    def is_openssh_ca(self) -> bool:
+        return SshHostCaExtension() in self.extensions or SshUserCaExtension() in self.extensions
```
Same as above, please do not write this in a single line.

mathiasertl commented on ca/django_ca/openssh/extensions.py:
```diff
@@ -0,0 +1,62 @@
+# pylint: disable=missing-function-docstring; https://github.com/PyCQA/pylint/issues/3605
+
+"""Module providing extensions for x509 authorities supporting OpenSSH.
+"""
```
stray newline here

thinkwelltwd replied:
Removed it

alfonsrv commented:
I think the use-case here is to use it with tools like [smallstep's CLI](https://github.com/smallstep/cli) to enable single-sign on for SSH.  

[Smallstep's CA](https://github.com/smallstep/certificates) seems to be very limited as in only supporting SSH certificates. Having the full capabilities django-ca offers paired with the workflow and security benefits of Step's SSO workflow for SSH would be a match made in heaven.

thinkwelltwd commented:
Thanks for your interest in this PR, @mathiasertl.  :grin: 

I thought you weren't interested, and so I moved on and implemented it myself in my own project, but I'd be glad to try to work with you further on this. I've made the changes you requested.

@alfonsrv is correct that I was trying to get a [smallstep-like CA](https://github.com/smallstep/certificates) for Python, where SSH keys are supported. A number of other API / design decisions will need to follow to get there. 

To implement that functionality further, I created the [aspen_ssh](https://github.com/thinkwelltwd/aspen_ssh) library which is based on the relevant pieces from the now-archived [Netflix Bless](https://github.com/Netflix/bless). This library is used to sign the SSH keys, and can be a dependency for django-ca, or else incorporated into django-ca

Would you like to merge this PR and then move on to other discussions and PRs?

mathiasertl commented on ca/django_ca/managers.py:
```diff
@@ -336,9 +342,21 @@ def init(
+        if openssh_ca:
+            algorithm = parse_hash_algorithm(algorithm)
+        else:
+            algorithm = None
```
I think that's a bug. This will force the algorithm to `None` if this is _not_ an OpenSSH CA. That's the inverse of what was there.

thinkwelltwd replied:
Gah nice catch! :man_facepalming:  Fixed now.

mathiasertl commented on ca/django_ca/managers.py:
```diff
@@ -336,9 +342,21 @@ def init(
+        if openssh_ca and parent:
+            raise ValueError("OpenSSH does not support intermediate authorities")
+        if not openssh_ca and key_type == "EdDSA":
+            raise ValueError("EdDSA only supported for OpenSSH authorities")
```
Q: Is that your restriction or a technical restriction?

thinkwelltwd replied:
http://cvsweb.openbsd.org/cgi-bin/cvsweb/src/usr.bin/ssh/PROTOCOL.certkeys?rev=1.8

> The valid key types for CA keys are ssh-rsa, ssh-dss and the ECDSA types
ecdsa-sha2-nistp256, ecdsa-sha2-nistp384, ecdsa-sha2-nistp521. "Chained"
certificates, where the signature key type is a certificate type itself
are NOT supported.

mathiasertl commented on ca/django_ca/openssh/__init__.py:
```diff
@@ -0,0 +1 @@
+from .extensions import *
```
Please no wildcard imports (flake8 rightfully complains about these). Also name exported extensions in `__all__`, as mypy will otherwise complain.

thinkwelltwd replied:
> Also name exported extensions in __all__, as mypy will otherwise complain.

```python
__all__ = (
    'SshHostCaType',
    'SshHostCaExtension',
    'SshUserCaType',
    'SshUserCaExtension',
)
```

I guess I'm a little confused. Are you saying you want the imports enumerated in both the module and the __init__.py file?

thinkwelltwd replied:
@mathiasertl could you clarify this comment:

> Please no wildcard imports (flake8 rightfully complains about these). Also name exported extensions in __all__, as mypy will otherwise complain.

I'm not sure what you want me to change in the `__all__` statement that's already there. Regarding the star import, I'm willing to enumerate the imports in the `__init__` file as well, although I don't understand why the namespace is polluted in this situation. But I'm not disputing your requests; only asking for clarification what you want. 

I'd also be glad if we could speed up the communications tempo.

mathiasertl replied:
With `from .extensions import ...` (whether wildcard or not), you effectively make extension classes implemented in `django_ca.openssh.extensions` available under the shorter and nicer `django_ca.openssh`. That's all well and good, no complaint about that in general.

You have a similar layout in `django_ca.extensions`. Try disabling one of the members in `__all__` in `django_ca.extensions.__init__` and run mypy against the source code. In this example, I've disabled the `KeyUsage` export:

```
$ mypy --strict ca/django_ca
ca/django_ca/models.py:82: error: Module "django_ca.extensions" does not explicitly export attribute "KeyUsage"; implicit reexport disabled  [attr-defined]
ca/django_ca/forms.py:31: error: Module "django_ca.extensions" does not explicitly export attribute "KeyUsage"; implicit reexport disabled  [attr-defined]
...
```

So that's why the `__all__` should be in `__init__.py`. (I believe pylint will also complain).

When you have a wildcard import, e.g. flake8 will not be able to catch some errors. Star imports are thus itself a flake8 warning, and any variable used from those star imports will also throw one.

thinkwelltwd replied:
Thanks for the explanation; I'm following you now, I believe. Does my latest commit match what you want?

mathiasertl commented:
The changes look good to me now! Thanks for helping to implement this.

Note that the test cases fail due to an unrelated change in the acme library. (and the failing test case is there precisely to detect such changes), so it does not merge blocking naturally.
